### DISCUSSION TITLE
MFRC522 released 🎉🎊

### DISCUSSION THREAD

**Discussion by salarizadi** [Moddable-OpenSource/moddable#1156] [Announcements]
Hello to all!

After 2 weeks of hard work, I was able to write this module, why did it take so long?

1. I wrote 3 libraries called RC522, which unfortunately did not work and I had to delete them completely.
2. The problem of lack of stack. Every time I tried to display and retrieve the card information in the ESP32, I got a Memory Abort error until I finally found a way to avoid this problem and quickly read the information. And there are really a lot of bytes that need to be put into arrays before they can be used
3. When I wanted to transfer data from XS to JavaScript, I had trouble transferring data until I was able to find a way to do this as well.
4. And...

Well, there are always challenges, but I enjoy these challenges!!!

[And now you can click on this link to download and use.](https://github.com/salarizadi/moddable-sdk/tree/main/Modules/MFRC522)

[If you like to support me, you can buy me coffee so I can make more modules](https://github.com/salarizadi/moddable-sdk/blob/main/README.md#buy-coffee-for-me-)

[If you want to see what happened during this time, you can read what @phoddie and I talked about](https://github.com/Moddable-OpenSource/moddable/discussions/1127)

**Comment by phoddie** [comment]
Congratulations on releasing this project. It looks like a huge amount of work! The API looks quite straightforward. I'm excited to try it out. First I need to get an MFRC522. ;)

**Reply by salarizadi** [reply]
I am glad that you want to test it, after testing, be sure to tell the problems of this module.

**Comment by tve** [comment]
Interesting! I haven't used NFC,,, Question: the chip seems to have SPI and I2C interfaces, why did you have to use native code in the driver?

**Reply by salarizadi** [reply]
i2c xs could meet expectations but spi should have been written.

io/spi xs doesn't quite live up to expectations, so I had to use native ESP IDF code, but in the future we'll definitely find a way to solve this problem written entirely in JavaScript.

But I tried to write a combination, for example, I completely entrusted the display of sectors to JavaScript.

**Reply by tve** [reply]
what does embedded:io/SPI not do that you needed? (naive question)

**Reply by salarizadi** [reply]
From this link you can see that some send and receive parameters are not supported by io/spi xs:

https://github.com/salarizadi/moddable-sdk/blob/main/Modules/MFRC522/MFRC522/esp32/MFRC522_Handler.c#L17


There are too many SPI configuration commands and too many different models to expect too much from io/spi xs.

**Reply by phoddie** [reply]
Interesting. SPI has many options and details. Like @tve, I'd like to understand what else might be needed. To be *very* clear, what you have done is impressive and great. I just can't help being a curious engineer about how we can improve.

ECMA-419 is intentionally designed to be extensible so we can support the features required. The link you provide to `PCD_WriteRegister` looks like it is just a synchronous SPI write. Maybe I am misreading the code? The ECMA-419 SPI write should do exactly the same. The `PCD_ReadRegister` call is a bit trickier because it does a write and read transaction. That is supported by ECMA-419 SPI with the `transfer()` method.

**Reply by salarizadi** [reply]
@phoddie 

Yes exactly there are many and varied commands, if you remember in the [MFRC522](https://github.com/Moddable-OpenSource/moddable/discussions/1127?sort=new#discussioncomment-6136797) module discussion I said that I was working on a separate library for SPI but I saw that I could write specifically for the MFRC522 module which luckily worked but I'm still working on it do . . in a separate SPI library and we will soon see the new IDF ESP SPI library, but it would be great if we could do it together for IO/SPI. This way it will be available to everyone under the original MODDABLE SDK.

**Reply by phoddie** [reply]
@salarizadi – I'm interested to see what you do. But it would be helpful if you be specific about what you believe is missing from the ECMA-419 SPI specification. That's sill unclear to me.
